### PR TITLE
Consolidate Store API and Core account creation handling

### PR DESCRIPTION
**Pull request by mikejolley**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Working on https://github.com/woocommerce/woocommerce/issues/50632 I found the need to duplicate logic between the Store API and other blocks. I was creating a utility when I noticed that the core functions were very similar to the helpers we created in blocks, therefore this PR consolidates them.

The reason they diverged in the first place was that blocks had some unique needs, but as a feature plugin could not update core directly. Now we can.

This PR will:

- Move legacy checkout specific handling from the customer creation functions to the legacy checkout classes
- Add missing docblocks to core from blocks
- Update wc_create_new_customer to ensure handling is generic. This includes things such as error messages.
- Remove handling for woocommerce_registration_generate_username/woocommerce_registration_generate_password from wc_create_new_customer so it works under all circumstances. Handling for those settings is found in the forms + checkout.
- Update Store API to use the core functions.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout as a guest via the shortcode checkout. 
2. Check the create account box.
3. Use an email that already has an account. Place order.
4. On legacy checkout the notice should read "An account is already registered with your email address. Please log in"
5. Use a new email. Confirm the account is created after placing the order.

6. As a logged out user, try to create an account via the my account page. 
7. Try invalid options first (existing account). Ensure error is shown.
8. Use a new email and create an account successfully.

9. Checkout as a guest via the block checkout. 
10. Check the create account box.
11. Use an email that already has an account. Place order.
12. Notice should read "An account is already registered with test@test.com. Please log in or use a different email address."
13. Use a new email. Confirm the account is created after placing the order.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/wc-user-functions.php [hunk 1]
```diff
@@ -55,17 +58,24 @@ function wc_create_new_customer( $email, $username = '', $password = '', $args =
...
+				sprintf(
+					// Translators: %s Email address.
+					esc_html__( 'An account is already registered with %s. Please log in or use a different email address.', 'woocommerce' ),
+					esc_html( $email )
+				)
+			);
 		}
 
-		if ( 'yes' === get_option( 'woocommerce_registration_generate_username', 'yes' ) && empty( $username ) ) {
+		if ( empty( $username ) ) {
```
Comment (senadir) [inline plugins/woocommerce/includes/wc-user-functions.php:71]: Do we need a dev note here? I think we can mention it in the release post.

File: plugins/woocommerce/includes/wc-user-functions.php [hunk 2]
```diff
@@ -74,35 +84,88 @@ function wc_create_new_customer( $email, $username = '', $password = '', $args =
 
 		// Handle password creation.
 		$password_generated = false;
-		if ( 'yes' === get_option( 'woocommerce_registration_generate_password' ) && empty( $password ) ) {
+
+		if ( empty( $password ) ) {
```
Comment (senadir) [inline plugins/woocommerce/includes/wc-user-functions.php:88]: Same here as above.

File: plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php [hunk 3]
```diff
@@ -588,55 +586,32 @@ private function get_request_payment_method( \WP_REST_Request $request ) {
...
-					);
-				case 'registration-error-empty-password':
-					throw new RouteException(
-						'registration-error-empty-password',
-						esc_html__( 'Please create a password for your account.', 'woocommerce' ),
-						400
-					);
-			}
+			// Set the customer auth cookie.
+			wc_set_customer_auth_cookie( $customer_id );
```
Comment (senadir) [inline plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:614]: Should this be gated behind a session check?

**Review by senadir — APPROVED** [review]
Looking good.

File: plugins/woocommerce/includes/wc-user-functions.php (hunk 1 above)
Comment (mikejolley) [reply plugins/woocommerce/includes/wc-user-functions.php:71]: Doesn't feel like a major change. Someone consuming the function would probably appreciate it being more predictable and independent of settings. We can cover with the release notes.

File: plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php (hunk 3 above)
Comment (mikejolley) [reply plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:614]: Is this not covered by the wrapper `should_create_customer_account`?

Comment (mikejolley) [reply plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:614]: Yeah, `should_create_customer_account` checks if they are logged in and returns false so this code only runs for guests. Same code as before actually.

**Pull request merged**